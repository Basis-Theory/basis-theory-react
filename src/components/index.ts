export * from './useBasisTheory';
export { useElements, Elements } from './Elements';
export * from './CardElement';
