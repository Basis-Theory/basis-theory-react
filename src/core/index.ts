export { useBasisTheory } from './useBasisTheory';
export { BasisTheoryProvider } from './BasisTheoryProvider';
export type { BasisTheoryReact } from './types';
export {
  BasisTheoryApiError,
  BasisTheoryValidationError,
} from '@basis-theory/basis-theory-js/common';
