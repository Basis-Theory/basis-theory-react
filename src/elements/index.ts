export { CardElement } from './CardElement';
export { TextElement } from './TextElement';
export { CardNumberElement } from './CardNumberElement';
export { CardExpirationDateElement } from './CardExpirationDateElement';
export { CardVerificationCodeElement } from './CardVerificationCodeElement';
