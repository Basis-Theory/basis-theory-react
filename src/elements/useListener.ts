import { useEffect } from 'react';
import type {
  BaseElement,
  ElementEventListener,
  EventType,
  Subscription,
} from '@basis-theory/basis-theory-js/types/elements';

const useListener = <
  T extends EventType,
  Ev,
  E extends BaseElement<unknown, Ev>
>(
  eventType: T,
  element?: E,
  listener?: ElementEventListener<Ev, T>
): void =>
  useEffect(() => {
    let subscription: Subscription;

    if (element && listener) {
      subscription = element.on(eventType, listener);
    }

    return () => subscription?.unsubscribe();
    // we shouldn't run this effect when eventType changes
    // which would turn out to be a different listener
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [element, listener]);

export { useListener };
