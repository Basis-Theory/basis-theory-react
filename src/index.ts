export * from './core';
export * from './elements';
