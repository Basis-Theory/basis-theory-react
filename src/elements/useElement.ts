import React, { useEffect, useRef, useState } from 'react';
import type {
  BaseElement,
  ElementType,
} from '@basis-theory/basis-theory-js/types/elements';
import { BasisTheoryReact } from '../core';
import { useBasisTheoryValue } from './useBasisTheoryValue';

const shallowDifference = <
  A extends Record<string, unknown>,
  B extends Record<string, unknown>
>(
  objA?: A,
  objB?: B
): Partial<A & B> =>
  [...new Set([...Object.keys(objA || {}), ...Object.keys(objB || {})])].reduce(
    (changed, key) => {
      const aValue = objA?.[key];
      const bValue = objB?.[key];

      if (aValue !== bValue) {
        return {
          ...changed,
          [key]: bValue,
        };
      }

      return changed;
    },
    {}
  );

/**
 * Creates, mounts and indexes an Element
 * with React lifecycle
 * @param id
 * @param type
 * @param options
 * @param wrapperRef
 * @param btFromProps
 * @returns created element and initial options used for mounting
 */
const useElement = <
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Element extends BaseElement<any, any>,
  Options extends unknown
>(
  id: string,
  type: ElementType,
  wrapperRef: React.RefObject<HTMLDivElement>,
  options: Options,
  btFromProps?: BasisTheoryReact
): Element | undefined => {
  const bt = useBasisTheoryValue(btFromProps);

  const [lastOptions, setLastOptions] = useState<Options>();

  const elementRef = useRef<Element | null>(null);

  useEffect(() => {
    if (bt && wrapperRef.current && !elementRef.current) {
      const newElement = bt.createElement(
        type as never,
        options as never
      ) as Element;

      elementRef.current = newElement;

      newElement.mount(`#${id}`).catch((mountError) => {
        setLastOptions(() => {
          throw mountError;
        });
      });
      bt.indexElement(id, newElement);
      setLastOptions(options);
    }

    // the only two dependencies that we need to watch for
    // are bt and wrapperRef. Anything else changing should not
    // be considered for creating/mounting an element
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [bt, wrapperRef]);

  const element = elementRef.current;

  useEffect(() => {
    if (element && options !== lastOptions) {
      const optionsDifference = shallowDifference(
        lastOptions as Record<string, unknown>,
        options as Record<string, unknown>
      );

      if (Object.keys(optionsDifference).length) {
        setLastOptions(options);
        element.update(optionsDifference).catch((updateError) => {
          setLastOptions(() => {
            throw updateError;
          });
        });
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [options, lastOptions]);

  return elementRef?.current || undefined;
};

export { useElement };
