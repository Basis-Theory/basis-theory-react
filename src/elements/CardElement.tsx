import React, { FC, useRef, ForwardedRef } from 'react';
import type {
  CardElement as ICardElement,
  CreateCardElementOptions,
  CardElementEvents,
  ElementEventListener,
  ElementStyle,
  CardElementValue,
} from '@basis-theory/basis-theory-js/types/elements';
import type { BasisTheoryReact } from '../core';
import { useElement } from './useElement';
import { useListener } from './useListener';

interface CardElementProps {
  id: string;
  bt?: BasisTheoryReact;
  style?: ElementStyle;
  disabled?: boolean;
  autoComplete?: 'on' | 'off';
  value?: CardElementValue<'static'>;
  onChange?: ElementEventListener<CardElementEvents, 'change'>;
  onFocus?: ElementEventListener<CardElementEvents, 'focus'>;
  onBlur?: ElementEventListener<CardElementEvents, 'blur'>;
  onReady?: ElementEventListener<CardElementEvents, 'ready'>;
  onKeyDown?: ElementEventListener<CardElementEvents, 'keydown'>;
  validateOnChange?: boolean;
  enableCopy?: boolean;
}

const CardElementC: FC<
  CardElementProps & { elementRef?: ForwardedRef<ICardElement> }
> = ({
  id,
  bt,
  style,
  disabled,
  autoComplete,
  value,
  onReady,
  onChange,
  onFocus,
  onBlur,
  onKeyDown,
  elementRef,
  validateOnChange,
  enableCopy,
}) => {
  const wrapperRef = useRef<HTMLDivElement>(null);
  const element = useElement<ICardElement, CreateCardElementOptions>(
    id,
    'card',
    wrapperRef,
    {
      enableCopy,
      validateOnChange,
      style,
      disabled,
      autoComplete,
      value,
    },
    bt,
    elementRef
  );

  useListener('ready', element, onReady);
  useListener('change', element, onChange);
  useListener('focus', element, onFocus);
  useListener('blur', element, onBlur);
  useListener('keydown', element, onKeyDown);

  return <div id={id} ref={wrapperRef} />;
};

export const CardElement = React.forwardRef<ICardElement, CardElementProps>(
  // eslint-disable-next-line get-off-my-lawn/prefer-arrow-functions
  function CardElement(props, ref) {
    return <CardElementC {...props} elementRef={ref} />;
  }
);

export type { CardElementProps };
