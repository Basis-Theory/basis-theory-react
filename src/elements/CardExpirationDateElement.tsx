import React, { FC } from 'react';
import type {
  CardExpirationDateElement as ICardExpirationDateElement,
  CreateCardExpirationDateElementOptions,
  ElementEventListener,
  ElementStyle,
  CardExpirationDateElementEvents,
} from '@basis-theory/basis-theory-js/types/elements';
import type { BasisTheoryReact } from '../core';
import { useElement } from './useElement';
import { useListener } from './useListener';

export interface CardExpirationDateElementProps {
  id: string;
  bt?: BasisTheoryReact;
  style?: ElementStyle;
  disabled?: boolean;
  'aria-label'?: string;
  placeholder?: string;
  onChange?: ElementEventListener<CardExpirationDateElementEvents, 'change'>;
  onFocus?: ElementEventListener<CardExpirationDateElementEvents, 'focus'>;
  onBlur?: ElementEventListener<CardExpirationDateElementEvents, 'blur'>;
  onReady?: ElementEventListener<CardExpirationDateElementEvents, 'ready'>;
  onKeyDown?: ElementEventListener<CardExpirationDateElementEvents, 'keydown'>;
}

export const CardExpirationDateElement: FC<CardExpirationDateElementProps> = ({
  id,
  bt,
  style,
  disabled,
  'aria-label': ariaLabel,
  placeholder,
  onReady,
  onChange,
  onFocus,
  onBlur,
  onKeyDown,
}) => {
  const element = useElement<
    ICardExpirationDateElement,
    CreateCardExpirationDateElementOptions
  >(
    id,
    'cardExpirationDate',
    {
      targetId: id,
      style,
      disabled,
      'aria-label': ariaLabel,
      placeholder,
    },
    bt
  );

  useListener('ready', element, onReady);
  useListener('change', element, onChange);
  useListener('focus', element, onFocus);
  useListener('blur', element, onBlur);
  useListener('keydown', element, onKeyDown);

  return <div id={id} />;
};
