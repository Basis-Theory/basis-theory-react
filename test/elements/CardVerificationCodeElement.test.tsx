import * as React from 'react';
import { CARD_BRANDS } from '@basis-theory/basis-theory-js/elements/constants';
import type {
  Brand,
  ElementStyle,
  CardVerificationCodeElement as ICardVerificationCodeElement,
} from '@basis-theory/basis-theory-js/types/elements';
import { render } from '@testing-library/react';
import { Chance } from 'chance';
import { CardVerificationCodeElement } from '../../src';
import { useElement } from '../../src/elements/useElement';
import { useListener } from '../../src/elements/useListener';

jest.mock('../../src/elements/useElement');
jest.mock('../../src/elements/useListener');

describe('CardVerificationCodeElement', () => {
  const chance = new Chance();
  const refArray = [React.createRef<ICardVerificationCodeElement>(), undefined];

  let id: string;
  let wrapperDiv: HTMLDivElement;
  let style: ElementStyle;
  let disabled: boolean;
  let autoComplete: 'on' | 'off';
  let ariaLabel: string;
  let placeholder: string;
  let cardBrand: Brand | string;
  let value: string;
  let validateOnChange: boolean;
  let onReady: jest.Mock;
  let onChange: jest.Mock;
  let onFocus: jest.Mock;
  let onBlur: jest.Mock;
  let onKeyDown: jest.Mock;
  let element: unknown;
  let ref: any;

  beforeEach(() => {
    id = 'my-card-verification-code';
    wrapperDiv = document.createElement('div');
    wrapperDiv.setAttribute('id', id);
    style = {
      [chance.string()]: chance.string(),
    };
    disabled = chance.bool();
    autoComplete = chance.pickone(['on', 'off']);
    ariaLabel = chance.string();
    placeholder = chance.string();
    value = chance.string();
    validateOnChange = chance.bool();
    cardBrand = chance.pickone<Brand>([...CARD_BRANDS]);
    onReady = jest.fn();
    onChange = jest.fn();
    onFocus = jest.fn();
    onBlur = jest.fn();
    onKeyDown = jest.fn();
    element = {
      [chance.string()]: chance.string(),
    };
    ref = chance.pickone(refArray);

    jest.mocked(useElement).mockReturnValue(element as any);
  });

  test('should match snapshot and call lifecycle hook properly', () => {
    const { container } = render(
      <CardVerificationCodeElement
        aria-label={ariaLabel}
        autoComplete={autoComplete}
        cardBrand={cardBrand}
        disabled={disabled}
        id={id}
        onBlur={onBlur}
        onChange={onChange}
        onFocus={onFocus}
        onKeyDown={onKeyDown}
        onReady={onReady}
        placeholder={placeholder}
        ref={ref}
        style={style}
        validateOnChange={validateOnChange}
        value={value}
      />
    );

    expect(container).toMatchSnapshot();
    expect(useElement).toHaveBeenCalledWith(
      id,
      'cardVerificationCode',
      { current: wrapperDiv },
      {
        targetId: id,
        style,
        disabled,
        autoComplete,
        'aria-label': ariaLabel,
        placeholder,
        cardBrand,
        value,
        validateOnChange,
      },
      undefined,
      // eslint-disable-next-line unicorn/no-null
      typeof ref === 'undefined' ? null : ref // undefined ref gets forwarded as null
    );
    expect(useListener).toHaveBeenCalledWith('ready', element, onReady);
    expect(useListener).toHaveBeenCalledWith('change', element, onChange);
    expect(useListener).toHaveBeenCalledWith('focus', element, onFocus);
    expect(useListener).toHaveBeenCalledWith('blur', element, onBlur);
    expect(useListener).toHaveBeenCalledWith('keydown', element, onKeyDown);
  });
});
