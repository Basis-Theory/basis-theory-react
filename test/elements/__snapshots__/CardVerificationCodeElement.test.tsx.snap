// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardVerificationCodeElement should match snapshot and call lifecycle hook properly 1`] = `
<div>
  <div
    id="my-card-verification-code"
  />
</div>
`;
